**What is Python Module?**

A Python module is a file containing Python definitions and statements. A module can define functions, classes, and variables. A module can also include runnable code. Grouping related code into a module makes the code easier to understand and use. It also makes the code logically organized.

**Python - OS Module**

It is possible to automatically perform many operating system tasks. The OS module in Python provides functions for creating and removing a directory (folder), fetching its contents, changing and identifying the current directory, etc.

You first need to import the os module to interact with the underlying operating system. So, import it using the import os statement before using its functions.

**Getting Current Working Directory**

In [ ]:
# getcwd() function confirms returns the current working directory
import os
os.getcwd()


'/content'

**Creating a Directory**

In [ ]:
os.mkdir("myAIproject")

**Changing the Current Working Directory**

In [ ]:
os.chdir("myAIproject") # changing current workign directory
os.getcwd()


'/content/myAIproject'

**Change CWD to Parent**

In [ ]:
#os.chdir("myAIproject")
#os.getcwd()

os.chdir("..")
os.getcwd()


'/content'

**Removing a Directory**

In [ ]:
os.rmdir("myAIproject")

However, you can not remove the current working directory. 

To remove it, you must change the current working directory, as shown below.

**List Files and Sub-directories**

In [ ]:
os.listdir("/content")

['.config', '.ipynb_checkpoints', 'myAIproject', 'sample_data']

In [ ]:
os.listdir()

['.config', 'sample_data']

**Python - Math Module**

**math module on Python docs**

**https://docs.python.org/3/library/math.html**

Some of the most popular mathematical functions are defined in the math module. These include trigonometric functions, representation functions, logarithmic functions, angle conversion functions, etc. In addition, two mathematical constants are also defined in this module.


Pi is a well-known mathematical constant, which is defined as the ratio of the circumference to the diameter of a circle and its value is 3.141592653589793.



In [ ]:
import math
math.pi

3.141592653589793

Another well-known mathematical constant defined in the math module is **e**. 
It is called **Euler's number** and it is a base of the natural logarithm. Its value is 2.718281828459045.

In [ ]:
math.e

The math module contains functions for calculating various trigonometric ratios for a given angle. The functions (sin, cos, tan, etc.) need the angle in radians as an argument.

We, on the other hand, are used to express the angle in degrees. 

The math module presents two angle conversion functions: degrees() and radians(), to convert the angle from degrees to radians and vice versa.

For example, the following statements convert the angle of 30 degrees to radians and back (Note: π radians is equivalent to 180 degrees).

In [ ]:
math.radians(30)

0.5235987755982988

In [ ]:
math.degrees(math.pi/6)

29.999999999999996

The following statements show sin, cos and tan ratios for the angle of 30 degrees (0.5235987755982988 radians):

In [ ]:
math.sin(0.5235987755982988)

0.49999999999999994

In [ ]:
math.cos(0.5235987755982988)

0.8660254037844387

In [ ]:
math.tan(0.5235987755982988)

0.5773502691896257

You may recall that sin(30)=0.5, cos(30)=32 (which is 0.8660254037844387) and tan(30)= 13 (which is 0.5773502691896257).

**math.log()**

The math.log() method returns the natural logarithm of a given number. The natural logarithm is calculated to the base e.

In [ ]:
math.log(10)

2.302585092994046

In [ ]:
math.log10(10)

1.0

**math.exp()**

The math.exp() method returns a float number after raising e to the power of the given number. In other words, exp(x) gives e**x

In [ ]:
math.exp(10)

22026.465794806718

**math.pow()**

The math.pow() method receives two float arguments, raises the first to the second and returns the result. In other words, pow(4,4) is equivalent to 4**4.

In [ ]:
math.pow(2,4)

16.0

**math.sqrt()**

The math.sqrt() method returns the square root of a given number.

In [ ]:
math.sqrt(100)

10.0

In [ ]:
math.sqrt(3)

1.7320508075688772

In [ ]:
# importing sqrt() and factorial from the
# module math
import math as mt

# if we simply do "import math", then
# math.sqrt(16) and math.factorial()
# are required.
print(mt.sqrt(16))
print(mt.factorial(6))


4.0
720


In [ ]:
# importing built in module datetime
import datetime
from datetime import date
import time
 
# Returns the number of seconds since the January 1st 1970
print(time.time()) 
 
# Converts a number of seconds to a date object
print(date.fromtimestamp(454554))

1675014899.810522
1970-01-06


**Representation Functions**

The ceil() function approximates the given number to the smallest integer, greater than or equal to the given floating point number. 

The floor() function returns the largest integer less than or equal to the given number.

In [ ]:
math.ceil(4.5867)

5

In [ ]:
math.floor(4.5687)

4

**Python - Statistics Module**

The statistics module provides functions to mathematical statistics of numeric data. The following popular statistical functions are defined in this module.

**Mean**

The **mean()** method calculates the arithmetic mean of the numbers in a list.

In [10]:
import statistics
statistics.mean([2,5,6,9])

5.5

In [11]:
import statistics as st 
lst = [2,5,6,9]
st.mean([2,5,6,9])

5.5

In [13]:
import statistics as st 
lst = [2,5,6,9,90,67]
st.mean(lst)

29.833333333333332

**Median**

The median() method returns the middle value of numeric data in a list.

In [14]:
statistics.median([1,2,3,8,9])

3

In [15]:
statistics.median([1,2,3,7,8,9])

5.0

In [16]:
statistics.median([1,2,3,8])

2.5

**Mode**

The mode() method returns the most common data point in the list.

In [20]:
statistics.mode([2,5,3,2,8,3,9,4,2,5,6,5,5,5,5,5,6,6,6,6,6,6,6,6])

6

**Standard Deviation**

The stdev() method calculates the standard deviation on a given sample in the form of a list.

In [21]:
statistics.stdev([1,1.5,2,2.5,3,3.5,4,4.5,5])

1.3693063937629153

**Learn about the statistics module in Python docs**
https://docs.python.org/3/library/statistics.html

**Python - Collections Module**

The collections module provides alternatives to built-in container data types such as list, tuple and dict.

**namedtuple()**

The namedtuple() function returns a tuple-like object with named fields. These field attributes are accessible by lookup as well as by index.

General usage of this function is:

collections.namedtuple(type_name, field-list)


The following statement declares a employee class having name, age and salary as fields.



**Declare a Named Tuple**

In [35]:
import collections as cl
employee = cl.namedtuple('employee', 'name, age, salary')

To create a new object of this namedtuple, do the following:

**Create Object of Named Tuple**

In [23]:
s1 = employee("Abhay", 32, 65000)

The values of the field can be accessible by attribute lookup:

In [36]:
s1.name

'Abhay'

In [37]:
s1.salary

65000

In [40]:
# Basic example
Points = collections.namedtuple('Point', ' x, y')
p = Points(x=34, y=45)     # instantiate with positional or keyword arguments
p[0] + p[1]                # indexable like the plain tuple (11, 22)



79

In [41]:
# unpack like a regular tuple
x, y = p 
x, y

(34, 45)

In [44]:
# fields also accessible by name
p.x - p.y               

-11

In [45]:
p                       # readable __repr__ with a name=value style

Point(x=34, y=45)

**OrderedDict()**

The OrderedDict() function is similar to a normal dictionary object in Python. However, it remembers the order of the keys in which they were first inserted.

In [30]:
import collections

d1 = collections.OrderedDict()
d1['A'] = 42
d1['C'] = 56
d1['B'] = 76
d1['D'] = 98

for k,v in d1.items():
    print (k,v)

A 42
C 56
B 76
D 98


Upon traversing the dictionary, pairs will appear in the order of their insertion.

**deque()**

A deque object support appends and pops from either ends of a list. It is more memory efficient than a normal list object. In a normal list object, the removal of any item causes all items to the right to be shifted towards left by one index. 
Hence, it is very slow.



In [46]:
q=collections.deque([5,30,180,1080])
q.appendleft(0)
q

deque([0, 5, 30, 180, 1080])

In [47]:
q.append(50)
q

deque([0, 5, 30, 180, 1080, 50])

In [48]:
q.pop()

q

deque([0, 5, 30, 180, 1080])

In [49]:
q.popleft()

q


deque([5, 30, 180, 1080])

**Random Module**

The random module is a built-in module to generate the pseudo-random variables. It can be used perform some action randomly such as to get a random number, selecting a random elements from a list, shuffle elements randomly, etc.

**Generate Random Floats**

The random.random() method returns a random float number between 0.0 to 1.0. The function doesn't need any arguments.

In [57]:
import random
random.random()

0.29265098741174933

**Generate Random Integers**

The random.randint() method returns a random integer between the specified integers.

In [80]:
random.randint(1, 1000)

972

In [81]:
random.randint(1, 100)

76

**Generate Random Numbers within Range**

The random.randrange() method returns a randomly selected element from the range created by the start, stop and step arguments. 
The value of start is 0 by default. Similarly, the value of step is 1 by default.



In [84]:
random.randrange(1, 10)

5

In [92]:
random.randrange(1, 10, 2)

3

In [96]:
random.randrange(0, 101, 10)

60

**Select Random Elements**

The random.choice() method returns a randomly selected element from a non-empty sequence. 

An empty sequence as argument raises an IndexError.

In [100]:
random.choice('computer')

'r'

In [102]:
random.choice([12,23,45,67,65,43])

45

In [103]:
random.choice((12,23,45,67,65,43))

65

**Shuffle Elements Randomly**

The random.shuffle() method randomly reorders the elements in a list.

In [106]:
numbers=[12,23,45,67,65,43]
random.shuffle(numbers)
numbers

[67, 12, 23, 43, 45, 65]

In [107]:
random.shuffle(numbers)
numbers

[65, 43, 45, 67, 12, 23]